### ISSUE TITLE
Inconsistency on the side effect of receiving a Follow activity

### ISSUE THREAD

**Issue by marnanel** [w3c-social/activitypump#320]
[Section 5.3](https://www.w3.org/TR/activitypub/#follow-activity-inbox) says:

> Every actor SHOULD have a followers collection. This is a list of everyone who has sent a Follow activity for the actor, added as a side effect. 

But [section 7.5](https://www.w3.org/TR/activitypub/#follow-activity-inbox) disagrees: the side effect is to send an Accept or a Reject activity. The actor is only added to the followers collection if the reply is Accept.

> The side effect of receiving [the Follow activity] in an inbox is that the server SHOULD generate either an Accept or Reject activity with the Follow as the object and deliver it to the actor of the Follow. [...] In the case of receiving an Accept referencing this Follow as the object, the server SHOULD add the actor to the object actor's Followers Collection. 

Also, I think "receiving" in the last sentence should be "sending".

**Comment by ap-socialhub** [comment]
This issue was moved to https://socialhub.activitypub.rocks/t/inconsistency-on-the-side-effect-of-receiving-a-follow-activity/207

**Comment by ap-socialhub** [comment]
This issue was moved to https://socialhub.activitypub.rocks/t/inconsistency-on-the-side-effect-of-receiving-a-follow-activity/216

**Comment by trwnh** [comment]
> In the case of receiving an Accept referencing this Follow as the object, the server SHOULD add the actor to the object actor's Followers Collection.

that last sentence seems backwards. it should likely indeed be "sending", but it should probably also say "add the [object actor] to the [actor's] Followers Collection". in other words, `Follow.actor` is added to `Accept.actor.followers` -- the person following you is added to your followers collection, not the other way around.

**Comment by evanp** [comment]
I agree on both points. We need to add two errata; one for the misleading and inaccurate description of `followers`, and the other for the inaccurate processing of `Accept` activities.

**Comment by evanp** [comment]
OK, there are 3 things to do here:

1. Change the erroneous explanation of `followers`.
2. Remove the sentence about receiving an `Accept` activity from the section on the `Follow` activity.
3. Add sections on `Accept` and `Reject` to section 6 on the ActivityPub API. The sentence that was removed may make sense there.

I am going to add errata for the first two. I think the third is important enough that we should address it in the next version of the spec, but not enough of an error to include in Errata. However, it would probably be helpful to cover in the Primer.

**Comment by evanp** [comment]
I've added an erratum for section 5.3 here: https://www.w3.org/wiki/ActivityPub_errata/Proposed

> Section 5.3 should begin, *Every actor SHOULD have a followers collection. This is where one would find a list of all the actors that are following the actor.*

**Comment by evanp** [comment]
I've added an erratum for section 7.5 "Follow Activity" here: https://www.w3.org/wiki/ActivityPub_errata/Proposed

> In section 7.5 "Follow Activity", the second paragraph is incorrect, and should be removed.

**Comment by evanp** [comment]
The third issue is called out specifically in #489 , so I think we remove the Needs Primer Page from this issue and leave it on that one.

**Comment by evanp** [comment]
I sent out a call for consensus on the SocialCG mailing list for these two errata.

**Comment by evanp** [comment]
It's been 2 weeks, no object to the CFC, so I think we can safely apply this erratum.

https://lists.w3.org/Archives/Public/public-swicg/2025May/0013.html

**Comment by gobengo** [comment]
per op

> But [section 7.5](https://www.w3.org/TR/activitypub/#follow-activity-inbox) disagrees: the side effect is to send an Accept or a Reject activity. The actor is only added to the followers collection if the reply is Accept.

No, it does not say that the side effect of a follow is [always] to send an accept. It recommends sending one, and that means it's not required for a server to generate Accept or listen for Accept.

There is a requirement to respond to a Reject, which Evan's proposing to remove by proposing to remove a paragraph from section 7.5? The erratum doesn't explain why you are removing that requirement or what error it fixes.

**Comment by trwnh** [comment]
The addition to the `followers`/`following` collection is the side effect of the Accept, not a side effect of the Follow. See https://github.com/w3c/activitypub/issues/489 for more.

**Comment by evanp** [comment]
We had an objection on the mailing list after I'd applied the errata (Thanks @gobengo !) so I rolled back the changes. We can move this back to "Needs errata" and we'll do another CFC once we have some consensus on the language.

**Comment by evanp** [comment]
I think Ben's primary objection is to follow the [W3C process for errata](https://www.w3.org/policies/process/#errata). In particular, I think he'd like us in the Errata document itself to describe the error in the document as published as well as the candidate correction, called out as such. It seems like we have some latitude in how to do this; one option might be just to name and link to originating the GitHub issue. Another would be to include a brief (hopefully) description of the error. I'll try to revisit this in future errata proposals.

**Comment by trwnh** [comment]
Perhaps we can use the following error description and candidate correction:

> **Error description**: In Section 7.5 "Follow activity", the second paragraph incorrectly states that receiving an Accept/Reject Follow will modify your followERS collection, when in fact it will modify your followING collection. Furthermore, the described side effects do not apply to a Follow; they apply to Accept/Reject Follow, and instead belong in Sections 7.6 "Accept activity" and 7.7 "Reject activity". (The requirements are already present and correct in those two sections.)
>
> **Candidate correction**: The second paragraph of Section 7.5 "Follow activity" should be removed.

**Comment by gobengo** [comment]
@trwnh Thanks for contributing an error description.

> Follow will modify your followERS collection, when in fact it will modify your followING collection.

Have you considered that it's often the case that the follower server also is storing a representation of the followed actors followers collection (usually as a cache, but it also could be for harassment)? This requirement, unchanged, perhaps serves some utility to require the attempted-follower's server not to include the non-follower in the followed actors 'followers' collection iff the followed actor rejected the follow of the attempted-follower.

>  Furthermore, the described side effects do not apply to a Follow; they apply to Accept/Reject Follow, and instead belong in Sections 7.6

This seems to imply that there is no way to appear in the following collection without there being an `Accept` or `Reject` activity, and I don't see how that could be the case. 

The `following` collection should be modified by activities sent to the `outbox`. (see section 6.5).
It is not only modified by receiving accepts/rejects in an inbox.

AFAICT An actor is in charge of declaring who it is `following`.
`Accept/Reject` is irrelevant.

And an Actor is also in charge of declaring who its `followers` are.
Someone who tried to follow someone else, but got a rejection, MUST not list themselves in the followers collection attributed to the actor that rejected their follow (e.g. when their server serves the other actor's followers collection, e.g. as a cache).

---

>  It seems like we have some latitude in how to do this; one option might be just to name and link to originating the GitHub issue. Another would be to include a brief (hopefully) description of the error. I'll try to revisit this in future errata proposals.

I agree that the requirement for an error description could be satisfied in a number of ways.

It's important to me that understanding the ActivityPub specification and errata does not rely on trawling through issue threads on another website like GitHub, especially considering that this website allows for later deletion/editing of the comments such that the erratum becomes ambiguous or even changes its meaning entirely. There is also already a track record of this website being bought and sold, content changed, TOS adjusted in big ways, as well as this website blocking access to users that are allowed to participate in w3.org discussions.

For these reasons I will advocate against any error description that includes a link to GitHub instead of a plain language description of the error that is meant to be internally consistent and interpreted without reference to an issue.

---

One more thing: It occurs to me as extremely risky to this whole spec to go and propose candidate corrections while just considering a single issue at a time and not the whole set of agreed upon errors. It may seem like nothing to you, but every little change creates hours of work for every implementation and testing tool, not to mention people like me legitimately trying to prevent accidental worsenings of the spec like already happened with the proposals yall made on the inbox definition for #289.

As I've asked before, *please* consider making your next erratum proposals *only* with an error description that we can come to consensus on. They are much less risky proposals and cheaper to assess. There is **no reason** you need to be including candidate corrections in the same proposals as an error description, and there is no reason we need to agree to your proposals to do so. Let's agree on what the errors even are, then we will have the context required to make the best candidate corrections.

**Comment by trwnh** [comment]
@gobengo 

> This requirement, unchanged, perhaps serves some utility to require the attempted-follower's server not to include the non-follower in the followed actors 'followers' collection iff the followed actor rejected the follow of the attempted-follower.

I don't think that makes sense at all. The way it's currently (erroneously) phrased is as follows:

> In the case of receiving an Accept referencing this Follow as the object, the server SHOULD add the actor to the object actor's [Followers Collection](https://www.w3.org/TR/activitypub/#followers)

This means the flow is like so:

- Alice Follow Bob
- Bob Accept(Follow) to:Alice
- Accept Follow received by Alice's server

Now, the text says that, since Alice sent the Follow, and since she got back an Accept from Bob, that Alice's server SHOULD add Bob to Alice's `followers`. Think about that for a second. Bob accepted that Alice follows Bob, so why is Alice's server behaving as if Bob now follows Alice instead? The correct and intended behavior is that Alice is now a follower of Bob. It is obviously an error to recommend that Alice's server should add Bob to Alice's followers collection.

It also remains true that the side effect being described is a consequence of the Accept, not of the Follow. And again, we can see that the Accept activity describes (correctly this time):

> If the object of an Accept received to an inbox is a Follow activity previously sent by the receiver, the server SHOULD add the actor to the receiver's [Following Collection](https://www.w3.org/TR/activitypub/#following).

So we have correct text in the correct section, and incorrect text in the incorrect section.

---

> This seems to imply that there is no way to appear in the following collection without there being an Accept or Reject activity

Correct. Following collections should not be modified outside of the Accept/Reject flow. When you send a Follow, you wait for an Accept before modifying the `followING` collection.

> AFAICT An actor is in charge of declaring who it is following

If you serve a `following` collection containing actors who have not Accepted your Follow yet, then you are making an untrusted claim[^1]. A well-behaved server managing your `following` collection for you would wait for an Accept.

[^1]: Honestly, the whole Follow flow makes it really hard to make trusted claims about follow relationships. You could *maybe* devise some kind of cross-signing scheme where both the Follow.actor and Follow.object sign for the Follow (or the Accept.actor and Accept.object.actor sign for the Accept), but that's kind of complicated. This combined with other difficulties around managing follow relationships (including destroying follow relationships) makes me wonder if an explicitly reified Relationship or Subscription object wouldn't have made more sense.

> The following collection should be modified by activities sent to the outbox. (see section 6.5).

It looks like Section 6.5 is committing the same class of error here, which is that the management of the `followING` collection is not a side effect of issuing a Follow, but rather it is a side effect of receiving the Accept. I'm not sure what a candidate correction here would look like, though. We do still have #489 for which we would need to add sections for C2S Accept and C2S Reject. Probably the end result is to have C2S Follow be very brief, while C2S Accept and C2S Reject describe the side effects of accepting or rejecting a follow.

---

> It may seem like nothing to you, but every little change creates hours of work for every implementation and testing tool, not to mention people like me legitimately trying to prevent accidental worsenings of the spec like already happened with the proposals yall made on the inbox definition for https://github.com/w3c/activitypub/issues/289.

I'm not sure what you mean specifically by "accidental worsenings" there, but no requirements were changed; only confusing implications were clarified. What meaningfully got worse by removing "a reference to"? What practical difference does this clarification of wording cause for anyone, if the requirements are still the same? I appreciate the watchfulness and additional attention being given to these proposed errata, but if conformance is unaffected then we're just wordsmithing. The normative text here makes it clear that you should wait for an Accept. It has always been the side effect of the Accept, not the Follow. Perhaps we still disagree on the error, but I feel like the error description I provided in my prior comment is sufficiently precise.

**Comment by evanp** [comment]
I'll update the proposed errata to reference the error.

**Comment by evanp** [comment]
I updated the erratum for section 5.3 to this text:

> Section 5.3 "Followers Collection" erroneously says that the "followers" collection is ''a list of '''everyone''' who has sent a Follow activity for the actor'' (emphasis added). Because the receiving actor can optionally Reject a Follow activity, the sending actor can later Undo a Follow activity, and one or both could Block the other, not everyone who sends a Follow activity will be in the list of followers.  A possible correction is to replace the text with ''Every actor SHOULD have a followers collection. This is where one would find a list of all the actors that are following the actor.''

**Comment by evanp** [comment]
I updated the erratum for section 7.5 to this text:

> In section 7.5 "Follow Activity", the second paragraph erroneously describes processing requirements for receiving an Accept activity with a Follow activity as the object. It describes changing the "followers" collection; however, receiving an "Accept" activity over the federation protocol would result in a change to the "following" collection. The correct behaviour is already described in section 7.6 "Accept Activity". One possible correction is to simply remove the second paragraph of section 7.5.

**Comment by evanp** [comment]
All right. I've taken off the needs errata label, and marked it as needs group input.  @gobengo if these new errata meet your needs, I can start a new CFC next week.

**Comment by evanp** [comment]
I sent off a CFC today. The mailing list thread starts here: https://lists.w3.org/Archives/Public/public-swicg/2025Aug/0031.html

**Comment by cristianolongoodhl** [comment]
I think that the followers collection remains underspecified, but I agree that the proposed modification would however improve the document.

**Comment by evanp** [comment]
The CFC was only for the 7.5 erratum, so I'm going to apply that now. I will also make a post for the 5.3 erratum so we can start the clock on that one too.

**Comment by evanp** [comment]
I added the erratum for 7.5 and applied it to the editor's draft. I also posted the 5.3 erratum for CFC: 

https://lists.w3.org/Archives/Public/public-swicg/2025Sep/0044.html

**Comment by evanp** [comment]
All right! Added to errata, and added to editor's draft.
